interfaces/builtin/modem_manager.go

```diff
@@ -0,0 +1,1205 @@
+
+dbus (send)
+   bus=system
+   path=/org/freedesktop/DBus
+   interface=org.freedesktop.DBus
+   member={Request,Release}Name
+   peer=(name=org.freedesktop.DBus),
+
+# Allow binding the service to the requested connection name
+dbus (bind)
+    bus=system
+    name="org.freedesktop.ModemManager1",
+
+# Allow traffic to/from our path and interface with any method
+dbus (receive, send)
+    bus=system
+    path=/org/freedesktop/ModemManager1{,/**}
+    interface=org.freedesktop.ModemManager1*,
+
+# Allow traffic to/from org.freedesktop.DBus for ModemManager service
+dbus (receive, send)
+    bus=system
+    path=/org/freedesktop/ModemManager1{,/**}
+    interface=org.freedesktop.DBus.*,
+`)
+
+var modemManagerConnectedPlugAppArmor = []byte(`
+# Description: Allow using ModemManager service. Reserved because this gives
+#  privileged access to the ModemManager service.
+# Usage: reserved
+
+#include <abstractions/dbus-strict>
+
+# Allow all access to ModemManager service
+dbus (receive, send)
+    bus=system
+    path=/org/freedesktop/ModemManager1{,/**}
+    peer=(label=###SLOT_SECURITY_TAGS###),
+`)
+
+var modemManagerPermanentSlotSecComp = []byte(`
+# Description: Allow operating as the ModemManager service. Reserved because this
+#  gives privileged access to the system.
+# Usage: reserved
+accept
+accept4
+bind
+connect
+getpeername
+getsockname
+getsockopt
+listen
+recv
+recvfrom
+recvmmsg
+recvmsg
+send
+sendmmsg
+sendmsg
+sendto
+setsockopt
+shutdown
+socketpair
+socket
+`)
+
+var modemManagerConnectedPlugSecComp = []byte(`
+# Description: Allow using ModemManager service. Reserved because this gives
+#  privileged access to the ModemManager service.
+# Usage: reserved
+
+# Can communicate with DBus system service
+connect
+getsockname
+recv
+recvmsg
+recvfrom
+send
+sendto
+sendmsg
+socket
+`)
+
+var modemManagerPermanentSlotDBus = []byte(`
+<!-- DBus policy for ModemManager (upstream version 1.4.0)
+  - TODO change to polkit one when we enable it -->
+<policy context="default">
+    <allow send_destination="org.freedesktop.ModemManager1"/>
+</policy>
+
+<policy user="root">
+    <allow own="org.freedesktop.ModemManager1"/>
+</policy>
+`)
+
+var modemManagerPermanentSlotUdev = []byte(`
+# Concatenation of all ModemManager udev rules
+# do not edit this file, it will be overwritten on update
+
+ACTION!="add|change|move", GOTO="mm_cinterion_port_types_end"
+SUBSYSTEM!="tty", GOTO="mm_cinterion_port_types_end"
+ENV{ID_VENDOR_ID}!="1e2d", GOTO="mm_cinterion_port_types_end"
+
+SUBSYSTEMS=="usb", ATTRS{bInterfaceNumber}=="?*", ENV{.MM_USBIFNUM}="$attr{bInterfaceNumber}"
+
+ATTRS{idVendor}=="1e2d", ATTRS{idProduct}=="0053", ENV{.MM_USBIFNUM}=="01", ENV{ID_MM_CINTERION_PORT_TYPE_GPS}="1"
+
+LABEL="mm_cinterion_port_types_end"
+# do not edit this file, it will be overwritten on update
+
+ACTION!="add|change|move", GOTO="mm_mbm_end"
+SUBSYSTEMS=="usb", GOTO="mm_mbm_check"
+GOTO="mm_mbm_end"
+
+LABEL="mm_mbm_check"
+
+# Ericsson F3507g
```

Comment:
Note that this device detection can be done in code from the snap too. For instance, ofono has two ways of detecting the modem type: the old one by using udev rules, and the new one, where it looks directly for vendor/product ID in udev events instead of looking for labels added by udev rules. It just happens that attaching labels used to be more convenient.

As I understand this, interfaces should be kind of general so this interface should be ideally reusable by programs that handle modems, say ModemManager AND ofono. So the ideal thing would be to put this application specific stuff (because these rules are MM rules, ofono has different ones - note again that only for a few modems) inside the snap, not in the gadget snap and not in the interface. As the snap should not add udev rules, this should be handled in code instead, as ofono does. No need for more permissions provided for the interface, as it is just a matter of listening to udev events.

However, this would require intensive modifications in ModemManager that we cannot assume. In the future world, if we get upstream buy-in, they would handle this. Meanwhile, I think this can be added to the interface, otherwise it would need to be added to every gadget snap, for which we do not necessarily will have control in the future. So somebody would install MM snap from the store and will find it is not able to detect the system modem.